base/pcre.jl

```diff
@@ -72,83 +78,68 @@ function info{T}(
-    if re_ptr == C_NULL
-        error("$(bytestring(errstr[1]))",
-              " at position $(erroff[1]+1)",
-              " in $(repr(pattern))")
-    end
-
+    errno = Ref{Int32}(0)
+    erroff = Ref{UInt32}(0)
+    re_ptr = ccall((:pcre2_compile_8, PCRE_LIB), Ptr{Void},
+                    (Cstring, UInt32, UInt32, Ref{Int32}, Ref{UInt32}, Ptr{Void}),
+    pattern, sizeof(pattern), options, errno, erroff, C_NULL)
+    re_ptr == C_NULL && error("PCRE compilation error: $(err_message(errno[])) at offset $(erroff[])")
     re_ptr
 end
 
-function study(regex::Ptr{Void}, options::Integer)
-    # NOTE: options should always be zero in current PCRE
-    errstr = Array(Ptr{UInt8},1)
-    errstr[1] = C_NULL
-    extra = ccall((:pcre_study, :libpcre), Ptr{Void},
-                  (Ptr{Void}, Int32, Ptr{Ptr{UInt8}}),
-                  regex, options, errstr)
-    if errstr[1] != C_NULL
-        error("$(bytestring(errstr[1]))")
-    end
-
-    ccall((:pcre_assign_jit_stack, :libpcre), Void,
-          (Ptr{Void}, Ptr{Void}, Ptr{Void}),
-          extra, C_NULL, JIT_STACK)
-    extra
+function jit_compile(regex::Ptr{Void})
+    errno = ccall((:pcre2_jit_compile_8, PCRE_LIB), UInt32,
+                  (Ptr{Void}, Int32),
+                  regex, JIT_COMPLETE)
+    errno == 0 || error("PCRE JIT error: $(err_message(errno))")
 end
 
-study(re::Ptr{Void}) = study(re, Int32(0))
+free_match_data(match_data) =
+    ccall((:pcre2_match_data_free_8, PCRE_LIB), Void, (Ptr{Void},), match_data)
+
+free_re(re) =
+    ccall((:pcre2_code_free_8, PCRE_LIB), Void, (Ptr{Void},), re)
+
+free_jit_stack(stack) =
+    ccall((:pcre2_jit_stack_free_8, PCRE_LIB), Void, (Ptr{Void},), stack)
 
-free_study(extra::Ptr{Void}) =
-    ccall((:pcre_free_study, :libpcre), Void, (Ptr{Void},), extra)
-free(regex::Ptr{Void}) =
-    ccall(unsafe_load(cglobal((:pcre_free, :libpcre),Ptr{Void})), Void, (Ptr{Void},), regex)
+free_match_context(context) =
+    ccall((:pcre2_match_context_free_8, PCRE_LIB), Void, (Ptr{Void},), context)
 
-function exec(regex::Ptr{Void}, extra::Ptr{Void}, str::SubString, offset::Integer,
-              options::Integer, ovec::Vector{Int32})
-    return exec(regex, extra, str.string, str.offset, offset, sizeof(str),
-                options, ovec)
+function err_message(errno)
+  buffer = Array(UInt8, 256)
+  ccall((:pcre2_get_error_message_8, PCRE_LIB), Void,
+        (Int32, Ptr{UInt8}, UInt32), errno, buffer, sizeof(buffer))
+  bytestring(pointer(buffer))
 end
 
-function exec(regex::Ptr{Void}, extra::Ptr{Void}, str::ByteString, offset::Integer,
-              options::Integer, ovec::Vector{Int32})
-    return exec(regex, extra, str, 0, offset, sizeof(str), options, ovec)
+function exec(re,subject,offset,options,match_data)
+    rc = ccall((:pcre2_match_8, PCRE_LIB), Cint,
+               (Ptr{Void}, Cstring, Csize_t, Csize_t, Cuint, Ptr{Void}, Ptr{Void}),
+               re, subject, sizeof(subject), offset, options, match_data, MATCH_CONTEXT)
```

Comment:
Ok, this is subtle but I think that using `Cstring` here may be inappropriate since we're a) passing the length of the data as another argument, rather than relying on NUL-termination of the data and b) apparently sometimes passing data that contains NUL bytes? cc: @stevengj